jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -130,15 +132,19 @@ public HttpChannelState(ConnectionMetaData connectionMetaData)
     }
 
     @Override
-    public void init()
+    public void initialize()
     {
-        _complianceViolationListener = getHttpConfiguration().getCombinedComplianceViolationListener().initialize();
-    }
+        List<ComplianceViolation.Listener> listeners = _connectionMetaData.getHttpConfiguration().getComplianceViolationListeners();
 
-    @Override
-    public Components getComponents()
-    {
-        return this;
+        ComplianceViolation.Listener listener;
+        if (listeners.isEmpty())
+            listener = ComplianceViolation.Listener.NOOP;
+        else if (listeners.size() == 1)
+            listener = listeners.get(0);
+        else
+            listener = new CompositeComplianceViolationListener(listeners);
+
+        _complianceViolationListener = listener.initialize();
```

Comment:
If there is a capturing listener, then this makes a composite, only to immediately replace it with an initialized composite.

I will see if we can do something better by initializing the list first and then building the composite.... stand by...

Reply:
I still don't love the allocation on the composite on every request.... but it is rare to have more than 1 listener.